Editorial review: Document the Summarizer API (#39263)
mdn/content: files/en-us/web/api/summarizer_api/using/index.md

---
title: Using the Summarizer API
slug: Web/API/Summarizer_API/Using
page-type: guide
---

{{DefaultAPISidebar("Summarizer API")}}

The [Summarizer API](/en-US/docs/Web/API/Summarizer_API) provides an asynchronous ({{jsxref("Promise")}}-based) mechanism for a website to feed a body of text into the browser's own internal language model and request that it returns a summary of the text based on specified options. This article explains how to use the fundamentals of the Summarizer API.

## Creating a summarizer

All of the Summarizer API's functionality is accessed through a single interface — {{domxref("Summarizer")}}.

> The first step in getting the browser LLM to output a summary is to create a `Summarizer` object instance. This is done using the {{domxref("Summarizer.create()")}} static method, which takes an options object as an argument specifying options for what kind of summary you want written:

Author: Elsewhere, I've used "AI model", so I've changed all instances to "AI model", and gotten rid of any mention of LLM. Is that OK?

> The first step in getting the browser LLM to output a summary is to create a `Summarizer` object instance. This is done using the {{domxref("Summarizer.create()")}} static method, which takes an options object as an argument specifying options for what kind of summary you want written:

Author: Yup, fixed.

> The first step in getting the browser LLM to output a summary is to create a `Summarizer` object instance. This is done using the {{domxref("Summarizer.create()")}} static method, which takes an options object as an argument specifying options for what kind of summary you want written:

Reviewer: SGTM!

```js
const summarizer = await Summarizer.create({
  sharedContext:
    "A general summary to help a user decide if the text is worth reading",
  type: "tl;dr",
  length: "short",
  format: "markdown",
  expectedInputLanguages: ["en-US"],
  outputLanguage: "en-US",
});
```

> The {{domxref("Summarizer.sharedContext", "sharedContext")}} provides a string that helps the LLM write a more suitable summary for the context the text is being used in, while the {{domxref("Summarizer.type", "type")}} specifies what kind of summary you want to provide (key bullet points? A "tl;dr" summary?).

Author: updated

We also specify the desired {{domxref("Summarizer.length", "length")}}, output {{domxref("Summarizer.format", "format")}}, {{domxref("Summarizer.expectedInputLanguages", "expectedInputLanguages")}}, and desired {{domxref("Summarizer.outputLanguage", "outputLanguage")}}. If the input and output languages are not specified, the language of the input text is auto-detected, and the output language will match the input language.

If the browser's LLM does not support the specified input or output languages, an error is thrown.

> [!NOTE]
> See the {{domxref("Summarizer.create", "create()")}} reference page for the full list of available options.

> EDITORIAL: What languages are supported by a given browser? Is it possible to add more? Does each specific implementation of the API support only a certain subset of languages, with the aim of adding more in the future?

Reviewer 2: In Chrome's implementation, there are 39 languages in total that the Translator API supports:

Arabic, Bangla, Bulgarian, Chinese, Croatian, Czech, Danish, Dutch, English, Finnish, French, German, Greek, Hebrew, Hindi, Hungarian, Indonesian, Italian, Japanese, Kannada, Korean, Lithuanian, Marathi, Norwegian, Polish, Portuguese, Romanian, Russian, Slovak, Slovenian, Spanish, Swedish, Tamil, Telugu, Thai, Traditional Chinese, Turkish, Ukrainian, Vietnamese.

Yes, we are hoping to expand this list. I believe the list could be different for different browsers.

Author: Thanks, @alexandrascript. I've added a note about the different language support and removed the editorial note for now. I suppose what I am most curious about is whether you can get the Summarizer API to return a different language summary of a given text — so, for example, if the text is in `en-US, can you get it to return a `fr` summary? Or is that not possible, and you would have to use the translator API, then the summarize API on the result?

I wasn't able to get the Summarizer API to return a different language summary in my tests — it always returned an error — but the fact that you can specify separate `expectedInputLanguages` and `outputLanguage` implies that you should be able to do so.

## Checking configuration support

> Before creating a `Summarizer`, you can check whether your desired configuration is supported by the current browser using the {{domxref("Summarizer.availability()")}} static method. For example:

Author: fixed

```js
const availability = await Summarizer.availability({
  type: "tl;dr",
  length: "short",
  format: "markdown",
  expectedInputLanguages: ["en-US"],
  outputLanguage: "en-US",
});
```

> This method returns an emumerated value indicating whether support is, or will be available for the specified set of options. The meaning of the `available` and `unavailable` values is fairly obvious, but there are also `downloadable` and `downloading` values that mean support for the specified set of options is available to download or currently downloading, respectively.

Author: good call; updated!

## Generating a summary

When you've determined that your desired configuration works and you've created a `Summarizer` instance, you can use it to generate a summary by calling the {{domxref("Summarizer.summarize()")}} instance method on it, passing it the text to summarize as an argument.

```js
const summary = await summarizer.summarize(myTextString);
console.log(summary);
```

It also optionally accepts an options object as a second argument that can accept a `context` string specific to this summary, and an abort {{domxref("AbortSignal", "signal")}} allowing the summary request to be aborted (see next section).

There is a streaming version of the `summarize()` method available — {{domxref("Summarizer.summarizeStreaming()")}} — that allows you to return the summary as a {{domxref("ReadableStream")}}:

```js

> const stream = await summarizer.summarizeStreaming(myTextString);

Author: fixed

const reader = stream.getReader();
let summary = "";

async function processText({ done, value }) {
  if (done) {
    console.log("Stream complete");
    console.log(summary);
    return;
  } else {
    summary += value;
    const result = await reader.read();
    processText(result);
  }
}

> const result = await reader.read();

Author: fixed

processText(result);
```

After a `Summarizer` instance has been created, you can remove it again using the {{domxref("Summarizer.destroy()")}} instance method.

## Cancelling summarize operations

You can cancel a pending `create()`, `summarize()`, or `summarizeStreaming()` operation using an {{domxref("AbortController")}}:

```js
const controller = new AbortController();
const summary = await summarizer.summarize(formData.get("summaryText"), {
  signal: controller.signal,
});

// ...

controller.abort();
```

> EDITORIAL: I've tried this, and it doesn't seem to work. Am I missing something?

Author: as above

```

```css hidden live-sample___summarizer-example
* {
  box-sizing: border-box;
}

html {
  font-family: Arial, Helvetica, sans-serif;
}

body {
  max-width: 600px;
  margin: 0 auto;
}

form div {
  display: flex;
  flex-direction: column;
  gap: 10px;
  margin-bottom: 20px;
}

select,
textarea,
.summary-output {
  padding: 5px;
}

.summary-output {
  min-height: 150px;
  border: 1px solid black;
  width: 100%;
  display: block;
}

.error {
  color: red;
}
```

Note that we won't show the CSS for this example, as none of it is relevant to understanding the Summarizer API.

### JavaScript

In our script, we start off by grabbing references to the `<form>`, `<textarea>`, submit `<button>`, summary output `<p>`, and two `<output>` elements.

```js live-sample___summarizer-example
const form = document.querySelector("form");
const textarea = document.querySelector("textarea");
const submitBtn = document.querySelector("button");

const summaryOutput = document.querySelector(".summary-output");
const inputCount = document.querySelector(".input-count");
const outputCount = document.querySelector(".output-count");
```

> Next, we use the {{domxref("EventTarget.when()")}} method to create {{domxref("Observable")}} objects representing:

Author: Yeah, good call. Updated. I just really like Observables ;-)

    "A general summary to help a user decide if the text is worth reading",
  type: "tl;dr",
  length: "short",
  format: "markdown",
  expectedInputLanguages: ["en-US"],
  outputLanguage: "en-US",
});
```

The {{domxref("Summarizer.sharedContext", "sharedContext")}} option provides a string that helps the AI model write a more suitable summary for the context the text is being used in, while the {{domxref("Summarizer.type", "type")}} specifies what kind of summary you want to provide, like key bullet points or a "tl;dr" summary.

We also specify the desired {{domxref("Summarizer.length", "length")}}, output {{domxref("Summarizer.format", "format")}}, {{domxref("Summarizer.expectedInputLanguages", "expectedInputLanguages")}}, and desired {{domxref("Summarizer.outputLanguage", "outputLanguage")}}. If the input and output languages are not specified, the language of the input text is auto-detected, and the output language will match the input language.

If the browser's AI model does not support the specified input or output languages, an error is thrown.

> [!NOTE]
> See the {{domxref("Summarizer.create_static", "create()")}} reference page for the full list of available options.

> Different implementations will likely support different languages.
> 

Reviewer 3: I wonder if it's better to leave this out for now?

```suggestion
```

## Checking configuration support

Before creating a `Summarizer`, you can check whether your desired configuration is supported by the current browser using the {{domxref("Summarizer.availability_static", "Summarizer.availability()")}} static method. For example:

```js
const availability = await Summarizer.availability({
  type: "tl;dr",
  length: "short",
  format: "markdown",
  expectedInputLanguages: ["en-US"],
  outputLanguage: "en-US",
});
```

> This method returns an emumerated value indicating whether support is, or will be available for the specified set of options:

Reviewer 3: ```suggestion
This method returns an enumerated value indicating whether support is, or will be available for the specified set of options:
```

- `downloadable` means that the implementation supports the requested options, but needs to download a model or some fine-tuning data.

> - `downloading` means that the implementation supports the requested options, but needs to finish an ongoing download.

Reviewer 3: Rewording suggestion above for consistency

Author: I think I've updated it as specified, but I wasn't 100% sure which suggestion you were referring to.

const summary = await summarizer.summarize(myTextString, {
  signal: controller.signal,
});

// ...

controller.abort();
```

## Monitoring download progress and usage

If the AI model for a particular summarizer is downloading (`availability()` returns `downloadable` and `downloading`), it is helpful to provide the user with feedback to tell them how long they need to wait before the operation completes.

The `Summarizer.create()` method can accept a `monitor` property, the value of which is a callback function that takes a {{domxref("CreateMonitor")}} instance as an argument. `CreateMonitor` has a {{domxref("CreateMonitor/downloadprogress_event", "downloadprogress")}} event available, which fires when progress is made on downloading the AI model.
You can use this event to expose loading progress data:

```js
const summarizer = await Summarizer.create({
  sharedContext:
    "A general summary to help a user decide if the text is worth reading",
  type: "tl;dr",
  length: "short",
  monitor: (monitor) => {
    monitor.addEventListener("downloadprogress", (e) => {
      console.log(`Downloaded ${Math.floor(e.loaded * 100)}%`);
    });
  },
});
```

> It is also worth mentioning that some implementations have an input quota that governs how many operations a website can request in a given period. The total quota can be accessed via the {{domxref("Summarizer.inputQuota")}} property, while the quota usage for a particular summary operation can be returned using the {{domxref("Summarizer.measureInputUsage()")}} method:

Reviewer 3: Not blocking, but what does "implementations" refer to?  We're talking about models, right? Or a combination of model + browser API?

```suggestion
Some implementations have an input quota that governs how many operations a website can request in a given period. The total quota can be accessed via the {{domxref("Summarizer.inputQuota")}} property, while the quota usage for a particular summary operation can be returned using the {{domxref("Summarizer.measureInputUsage()")}} method:
```

For example:

```js
// Return total available quota
console.log(summarizer.inputQuota);

// Return quota usage for summarizing this particular string
const usage = await summarizer.measureInputUsage(myTextString);
console.log(usage);
```

## Complete example

Let's look at a complete example that demonstrates the Summarizer API in action.

### HTML

> In our markup, we first define an input {{htmlelement("form")}} that allows the user to set the text to summarize, and configuration options. This includes a {{htmlelement("textarea")}} for entering the text itself, an {{htmlelement("output")}} element to display the text character count, and two {{htmlelement("select")}} elements for choosing a summarizer {{domxref("Summarizer.type", "type")}} and {{domxref("Summarizer.length", "length")}}.

Reviewer 3: ```suggestion
In our markup, we first define an input {{htmlelement("form")}} that allows the user to set the text to summarize, and configuration options. This includes a {{htmlelement("textarea")}} for entering the text to be summarized, an {{htmlelement("output")}} element to display the character count of the text set by the user, and two {{htmlelement("select")}} elements for choosing a summarizer {{domxref("Summarizer.type", "type")}} and {{domxref("Summarizer.length", "length")}}.
```

* {
  box-sizing: border-box;
}

html {
  font-family: Arial, Helvetica, sans-serif;
}

body {
  max-width: 600px;
  margin: 0 auto;
}

form div {
  display: flex;
  flex-direction: column;
  gap: 10px;
  margin-bottom: 20px;
}

select,
textarea,
.summary-output {
  padding: 5px;
}

.summary-output {
  min-height: 150px;
  border: 1px solid black;
  width: 100%;
  display: block;
}

.error {
  color: red;
}
```

Note that we won't show the CSS for this example, as none of it is relevant to understanding the Summarizer API.

### JavaScript

In our script, we start off by grabbing references to the `<form>`, `<textarea>`, submit `<button>`, summary output `<p>`, and two `<output>` elements.

```js live-sample___summarizer-example
const form = document.querySelector("form");
const textarea = document.querySelector("textarea");
const submitBtn = document.querySelector("button");

const summaryOutput = document.querySelector(".summary-output");
const inputCount = document.querySelector(".input-count");
const outputCount = document.querySelector(".output-count");
```

Next, we use the {{domxref("EventTarget.addEventListener()")}} method to listen to two sets of events:

> - `click` events on the `<button>` element; when the button is clicked, the `handleSubmission()` function is called (we're not doing a real form `submit`, as they are disabled in the MDN live example embeds for security purposes).

Reviewer 3: ```suggestion
- `click` events on the `<button>` element; when the button is clicked, the `handleSubmission()` function is called (we're not doing a real form `submit`).
```

You can do this with `"allow-forms"`, like:

```md
{{embedlivesample("example name", "400px", , , , , "allow-forms")}}
```

See https://developer.mozilla.org/en-US/docs/MDN/Writing_guidelines/Page_structures/Live_samples#sandbox

Author: Nice! I've updated the code and wording to use the real submit.

  } else if (formData.get("summaryText").length < 100) {
    summaryOutput.innerHTML = `<span class="error">I'm not trying to summarize something that short!</span>`;
    return;
  } else {
    summaryOutput.innerHTML = "";
  }

  try {
    const summarizer = await Summarizer.create({
      sharedContext:
        "A general summary to help a user decide if the text is worth reading",
      type: formData.get("summaryType"),
      length: formData.get("summaryLength"),
    });

    summaryOutput.textContent = "...generating summary...";
    outputCount.textContent = "Output summary character count: -";
    submitBtn.disabled = true;

    const summary = await summarizer.summarize(formData.get("summaryText"));

    summaryOutput.textContent = summary;
    displayOutputCount();
    submitBtn.disabled = false;
  } catch (e) {
    summaryOutput.innerHTML = `<span class="error">${e}</span>`;
  }
}
```

The final step is to call the `updateInputCount()` function at the top level of the script, to ensure that the first `<output>` element containing the input count always displays the correct value on page load.

```js live-sample___summarizer-example
updateInputCount();
```

### Result

> The rendered exaple looks like this:

Reviewer 3: ```suggestion
The rendered example looks like this:
```

> {{EmbedLiveSample("summarizer-example", "100%", "400px")}}

Reviewer 3: 750px is tall enough for me to get rid of scroll:

```suggestion
{{EmbedLiveSample("summarizer-example", "100%", "750px")}}
```

Reviewer 3: And like this for enabling forms, I believe:

```
{{EmbedLiveSample("summarizer-example", , "750px", , , , , "allow-forms")}}
```

> Try entering a body of text into the `<textarea>` then pressing the button to generate an AI-generated summary. The text from your favourite Wikipedia page would be ideal. Try generating multiple summaries with different option combinations, to see how they affect the output.

Reviewer 3: 🇬🇧 😢 

```suggestion
Try entering a body of text into the `<textarea>` then pressing the button to generate an AI-generated summary. The text from your favorite Wikipedia page would be ideal. Try generating multiple summaries with different option combinations, to see how they affect the output.
```

    summaryOutput.innerHTML = `<span class="error">I'm not trying to summarize something that short!</span>`;
    return;
  } else {
    summaryOutput.innerHTML = "";
  }

  try {
    const summarizer = await Summarizer.create({
      sharedContext:
        "A general summary to help a user decide if the text is worth reading",
      type: formData.get("summaryType"),
      length: formData.get("summaryLength"),
    });

    summaryOutput.textContent = "...generating summary...";
    outputCount.textContent = "Output summary character count: -";
    submitBtn.disabled = true;

    const summary = await summarizer.summarize(formData.get("summaryText"));

    summaryOutput.textContent = summary;
    displayOutputCount();
    submitBtn.disabled = false;
  } catch (e) {
    summaryOutput.innerHTML = `<span class="error">${e}</span>`;
  }
}
```

The final step is to call the `updateInputCount()` function at the top level of the script, to ensure that the first `<output>` element containing the input count always displays the correct value on page load.

```js live-sample___summarizer-example
updateInputCount();
```

### Result

The rendered example looks like this:

> {{EmbedLiveSample("summarizer-example", , "750px", , , , , "allow-forms")}}

Author: Oh, darn it, I'd not even thought about that. It has been working for me locally, but of course it would do.

Is there a way that we can set permissions policies for MDN live demos?

> {{EmbedLiveSample("summarizer-example", , "750px", , , , , "allow-forms")}}

Reviewer 3: Oh, I forgot we have this already:

https://developer.mozilla.org/en-US/docs/MDN/Writing_guidelines/Page_structures/Live_samples#allow

```suggestion
{{EmbedLiveSample("summarizer-example", , "750px", , , , "summarizer", "allow-forms")}}
```

> {{EmbedLiveSample("summarizer-example", , "750px", , , , , "allow-forms")}}

Reviewer 3: It's alive!
